subsys/bluetooth/host/classic/spp.c

```diff
@@ -0,0 +1,490 @@
+/*
+ * Copyright 2024 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/classic/rfcomm.h>
+#include <zephyr/bluetooth/classic/sdp.h>
+#include <zephyr/bluetooth/classic/spp.h>
+
+#include "host/hci_core.h"
+#include "host/conn_internal.h"
+
+#include "l2cap_br_internal.h"
+#include "rfcomm_internal.h"
+
+#define LOG_LEVEL CONFIG_BT_SPP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_spp);
+
+#define MAX_SPP_CHANNEL 4
+#define SPP_UUID128_LEN 16
+#define SPP_RFCOMM_MTU  CONFIG_BT_L2CAP_TX_MTU
+
+#define SPP_INDEX_TO_ID(x) ((x) | 0x80)
+#define SPP_ID_TO_INDEX(x) ((x) & (~0x80))
+
+#define SDP_CLIENT_BUF_LEN 512
+
+enum {
+	SPP_STATE_IDLE,
+	SPP_STATE_DISCOVERING,
+	SPP_STATE_USED,
+};
+
+struct bt_spp_server {
+	uint8_t spp_id;
+	uint8_t state: 4;
+	uint8_t active_connect: 1;
+	uint8_t spp_uuid[SPP_UUID128_LEN];
+
+	struct bt_conn *conn;
+	struct bt_rfcomm_server rfcomm_server;
+	struct bt_rfcomm_dlc rfcomm_dlc;
+};
+
+static K_MUTEX_DEFINE(spp_lock);
+static struct bt_spp_cb *spp_cb;
+static struct bt_spp_server spp_server[MAX_SPP_CHANNEL];
+static uint8_t spp_state;
+static struct bt_uuid_128 sdp_uuid;
+
+NET_BUF_POOL_FIXED_DEFINE(sdp_client_pool, CONFIG_BT_MAX_CONN, SDP_CLIENT_BUF_LEN, 8, NULL);
+NET_BUF_POOL_FIXED_DEFINE(spp_pool, CONFIG_BT_MAX_CONN, BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
+			  16, NULL);
+
+#define BT_SPP_ATTRS(i)                                                                            \
+	static struct bt_sdp_attribute spp_attrs##i[] = {                                          \
+		BT_SDP_NEW_SERVICE,                                                                \
+		BT_SDP_LIST(BT_SDP_ATTR_SVCLASS_ID_LIST, BT_SDP_TYPE_SIZE_VAR(BT_SDP_SEQ8, 17),    \
+			    BT_SDP_DATA_ELEM_LIST({BT_SDP_TYPE_SIZE(BT_SDP_UUID128),               \
+						   spp_server[i].spp_uuid}, )),                    \

```

Comment:
Whether is the SDP record used for general purpose? From SPP specific, the UUID should be UUID16 and value is `SerialPort` (`0x1101`).

Reply:
here we support port when register as spp server. someone would like to register with private uuid, I add add  new API to solve it 
`int bt_spp_register_srv(uint8_t port, const struct bt_uuid *uuid);
`

but it is hard to register 16/32/128 uuid with stack bt_sdp_attribute  register, i  have not idear now , but I  think   it could be resolved